src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/ValidateAnomalyDetectorActionHandler.java

```diff
@@ -0,0 +1,590 @@
+            anomalyDetectionIndices
+                .initAnomalyDetectorIndex(
+                    ActionListener.wrap(response -> onCreateMappingsResponse(response), exception -> onFailure(exception))
+                );
+        } else {
+            preDataValidationSteps();
+        }
+    }
+
+    public void preDataValidationSteps() {
+        List<String> missingFields = new ArrayList<>();
+        List<String> formatErrors = new ArrayList<>();
+        if (anomalyDetector.getName() == null || anomalyDetector.getName() == "") {
+            missingFields.add("name");
+        } else if (!anomalyDetector.getName().matches(NAME_REGEX)) {
+            formatErrors.add(anomalyDetector.getName());
+            failuresMap.put("format", formatErrors);
+        }
+        if (anomalyDetector.getTimeField() == null) {
+            missingFields.add("time_field");
+        }
+        if (anomalyDetector.getIndices() == null) {
+            missingFields.add("indices");
+        }
+        if (anomalyDetector.getWindowDelay() == null) {
+            missingFields.add("window_delay");
+        }
+        if (anomalyDetector.getDetectionInterval() == null) {
+            missingFields.add("detector_interval");
+        }
+        if (anomalyDetector.getFeatureAttributes().isEmpty()) {
+            missingFields.add("feature_attributes");
+        }
+        if (!missingFields.isEmpty()) {
+            failuresMap.put("missing", missingFields);
+        }
+        String error = RestHandlerUtils.validateAnomalyDetector(anomalyDetector, maxAnomalyFeatures);
+        if (StringUtils.isNotBlank(error)) {
+            List<String> dupErrorsFeatures = new ArrayList<>();
+            dupErrorsFeatures.addAll(Arrays.asList(error.split("\\r?\\n")));
+            failuresMap.put("duplicates", dupErrorsFeatures);
+        }
+        if (!failuresMap.isEmpty()) {
+            validateAnomalyDetectorResponse();
+        } else {
+            validateNumberOfDetectors();
+        }
+    }
+
+    public void validateNumberOfDetectors() {
+        try {
+            QueryBuilder query = QueryBuilders.matchAllQuery();
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder().query(query).size(0).timeout(requestTimeout);
+            SearchRequest searchRequest = new SearchRequest(ANOMALY_DETECTORS_INDEX).source(searchSourceBuilder);
+            client.search(searchRequest, ActionListener.wrap(response -> onSearchAdResponse(response), exception -> onFailure(exception)));
+        } catch (Exception e) {
+            onFailure(e);
+        }
+    }
+
+    private void onSearchAdResponse(SearchResponse response) throws IOException {
+        if (response.getHits().getTotalHits().value >= maxAnomalyDetectors) {
+            failuresMap
+                .computeIfAbsent("others", k -> new ArrayList<>())
+                .add("Can't create anomaly detector more than " + maxAnomalyDetectors);
+            validateAnomalyDetectorResponse();
```

Comment:
Instead of returning response if max_num is reached, can we continue validation, and add "Please delete unused detectors" as suggested change?

Reply:
fixed